spec/errors.md

```diff
@@ -241,82 +242,53 @@ or for private implementation use that is not supported by the current implement
 > * {{The value is not one.}}
 > ```
 
-### Invalid Expression
-
-An **_<dfn>Invalid Expression</dfn>_** error occurs when a _message_ includes an _expression_
-whose implementation-defined internal requirements produce an error during _function resolution_
-or when a _function_ returns a value (such as `null`) that the implementation does not support.
-
-An **_<dfn>Operand Mismatch Error</dfn>_** is an _Invalid Expression_ error that occurs when
-an _operand_ provided to a _function_ during _function resolution_ does not match one of the
-expected implementation-defined types for that function;
-or in which a literal _operand_ value does not have the required format
-and thus cannot be processed into one of the expected implementation-defined types
-for that specific _function_.
-
-> For example, the following _message_ produces an _Operand Mismatch Error_
-> (a type of _Invalid Expression_ error)
-> because the literal `|horse|` does not match the production `number-literal`,
-> which is a requirement of the function `:number` for its operand:
-> ```
-> .local $horse = {horse :number}
-> {{You have a {$horse}.}}
-> ```
-> The following _message_ might produce an _Invalid Expression_ error if the
-> the function `:function` threw an exception or otherwise emitted an error
-> rather than returning a valid value:
->```
-> {{This has an invalid expression {$var :function} because it has a bug in it.}}
->```
-
 ### Unsupported Statement
 
 An **_<dfn>Unsupported Statement</dfn>_** error occurs when a message includes a _reserved statement_.
 
 > For example, attempting to format this message
-> would always result in an _Unsupported Statement_ error:
+> would result in an _Unsupported Statement_ error:
 >
 > ```
 > .some {|horse|}
 > {{The message body}}
 > ```
 
-## Selection Errors
+### Bad Selector
 
-**_<dfn>Selection Errors</dfn>_** occur when message selection fails.
+A **_<dfn>Bad Selector</dfn>_** error occurs when a message includes a _selector_
+with a resolved value which does not support selection.
 
-> For example, attempting to format either of the following messages
-> might result in a _Selection Error_ if done within a context that
-> uses a `:number` selector function which requires its input to be numeric:
->
-> ```
-> .match {|horse| :number}
-> 1 {{The value is one.}}
-> * {{The value is not one.}}
-> ```
+> For example, attempting to format this message
+> would result in a _Bad Selector_ error:
 >
 > ```
-> .local $sel = {|horse| :number}
-> .match {$sel}
-> 1 {{The value is one.}}
-> * {{The value is not one.}}
+> .local $day = {|2024-05-01| :date}
+> .match {$day}
+> * {{The due date is {$day}}}
 > ```
 
-## Formatting Errors
+## Message Function Errors
```

Comment:
The header should be singular too, even though it encapsulates other errors

```suggestion
## Message Function Error
```

Reply:
All of the other titles at this level use "Errors".

Reply:
"Unsupported Statement", "Bad Selector", "Bad Operand", "Bad Option", and "Bad Variant Key" are all singular. I guess Syntax Error and Resolution Error are presented as plural. Either way, we probably should be consistent.

In this case, "Message Function Error" is both a _class_ of errors and an error in its own right.

Reply:
It's a class of errors, next to Syntax Errors, Data Model Errors, and Resolution Errors. Its naming is consistent with those, both in how they're titled here, and the way the name includes the word "Error". The spec language is in fact quite deliberately careful about not requiring a bare Message Function Error to be emittable; instead it either refers directly to a specific Message Function Error, or an "appropriate" Message Function Error.

Reply:
If it's a class of errors, shouldn't there be a concrete subtype for all of the known errors?

I think the problem here is that its vaguely clear that there is a generic Message Function Error error type.

Currently:
- Syntax Error is a class and a type (and that's fine).
- Data Model Error has an exhaustive list of error types.
- Resolution Error has an exhaustive list of error types.
- Message Function Error has a partial list, where the missing type appears to be the "💩 Happened Error" aka "implementation defined error"

Reply:
> If it's a class of errors, shouldn't there be a concrete subtype for all of the known errors?

No, I don't think we need to make this list exhaustive. The others are exhaustive as they're emitted from our own code, but we should not presume that we can imagine all the things that can happen, and all the things that might go wrong, with custom functions.

Reply:
Actually, I think I'm trying to say that. The "catch all" error is a message function error, since everything else that can go user-visibly wrong is accounted for.

Reply:
So this thread can be resolved and the title left as it currently is, matching the others which include "Syntax Errors"?